### PR TITLE
feat: add visitor to check closures with free variables

### LINKED ISSUE #100: Add visitor to check closures with free variables

### PR DESCRIPTION
**Pull request by LunaAmora**

### CODE REVIEW

File: crates/atiny-checker/src/check/expr.rs [hunk 1]
```diff
@@ -4,13 +4,12 @@ use super::Check;
 use crate::infer::Infer;
 use crate::{context::Ctx, types::*, unify::unify};
 
+use atiny_tree::elaborated::Elaborated;
 use atiny_tree::r#abstract::Expr;
 
-type Elaborated = atiny_tree::elaborated::Expr<Type>;
-
 impl Check<'_> for &Expr {
```
Comment (algebraic-sofia) [inline crates/atiny-checker/src/check/expr.rs:10]: Why &Expr and use self in the method implementation?

File: crates/atiny-error/src/lib.rs [hunk 2]
```diff
@@ -3,6 +3,39 @@ use std::fmt::Display;
 
 use atiny_location::{ByteRange, NodeId, Point, Range};
 
+pub trait ErrorCreation {
```
Comment (algebraic-sofia) [inline crates/atiny-error/src/lib.rs:6]: Cool trait

File: crates/atiny-location/src/lib.rs [hunk 3]
```diff
@@ -130,9 +126,9 @@ impl<T> Located<T> {
 }
 
 pub trait WithLoc: Sized {
-    fn with_loc<Any>(self, located: &Located<Any>) -> Located<Self> {
+    fn loc(self, location: ByteRange) -> Located<Self> {
```
Comment (algebraic-sofia) [inline crates/atiny-location/src/lib.rs:129]: Maybe located is a better name, loc says nothing about the meaning of the function

Comment (algebraic-sofia) [inline crates/atiny-location/src/lib.rs:129]: Maybe located is a better name, loc says nothing about the meaning of # the function

File: crates/atiny-tree/src/elaborated.rs [hunk 4]
```diff
@@ -32,27 +34,29 @@ impl<T> VariableNode<T> {
...
 
-    Abstraction(VecDeque<Labeled<T>>, Box<Self>, T),
-    Application(Box<Self>, Vec<Self>, T),
+    Abstraction(VecDeque<Labeled<T>>, Box<Elaborated<T>>, T),
+    Application(Box<Elaborated<T>>, Vec<Elaborated<T>>, T),
 
-    RecordCreation(Symbol, Vec<Labeled<Self>>),
-    RecordUpdate(Box<Self>, Vec<Labeled<Self>>),
-    RecordField(Symbol, Symbol, Box<Self>),
+    RecordCreation(Symbol, Vec<Labeled<Elaborated<T>>>),
```
Comment (algebraic-sofia) [inline crates/atiny-tree/src/elaborated.rs:57]: Instance is a better name for this sob sob, just being boring rn

**Review by algebraic-dev — APPROVED** [review]
Nice

**Review by algebraic-dev — APPROVED** [review]

**Pull request merged**